### ISSUE TITLE
pdbtool: g_error_free: assertion 'error != NULL' failed [Debian]

### ISSUE THREAD

**Issue by jeffsf** [syslog-ng/syslog-ng#4034]
# syslog-ng
jeff@deb-devel:~/devel/syslog-ng-pdb$ syslog-ng -V
syslog-ng 3 (3.28.1)
Config version: 3.22
Installer-Version: 3.28.1
Revision: 3.28.1-2+b1

_Edit: #4036 blocks building 3.36.1 from tarball for testing._

## Platform
jeff@deb-devel:~/devel/syslog-ng-pdb$ cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

## To Replicate

Copy the example with test data from https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/3.26/administration-guide/81#TOPIC-1431219 (confirmed to be the same example in the latest documentation.) and paste into `doc-example.xml`

Run `pdbtool test ./doc-example.xml`

Observe the error message

```
(pdbtool:3810073): GLib-CRITICAL **: 15:26:51.831: g_error_free: assertion 'error != NULL' failed
NOT Testing message as program is unset: message='Accepted password for sampleuser from 10.50.0.247 port 42156 ssh2'
```

Expected:
* `pdbtool` would run without error

Change the content of `doc-example.com` to provide the program

```
-                            <test_message>Accepted password for sampleuser from 10.50.0.247 port 42156 ssh2</test_message>
+                            <test_message program="ssh">Accepted password for sampleuser from 10.50.0.247 port 42156 ssh2</test_message>
```

Observe apparently the same error

```
(pdbtool:3818786): GLib-CRITICAL **: 15:29:51.113: g_error_free: assertion 'error != NULL' failed
Testing message: program='ssh' message='Accepted password for sampleuser from 10.50.0.247 port 42156 ssh2'
 Wrong match name='.classifier.rule_id', value='', expected='182437592347598'
 Wrong match name='SSH.AUTH_METHOD', value='', expected='password'
 Wrong match name='SSH_USERNAME', value='', expected='sampleuser'
 Wrong match name='SSH_CLIENT_ADDRESS', value='', expected='10.50.0.247'
 Wrong match name='SSH_PORT_NUMBER', value='', expected='42156'
```
 

_Edit: Confirming that this is not specific to the `test` mode_

```
$ pdbtool dump -p ./doc-example.xml -P ssh

(pdbtool:100226): GLib-CRITICAL **: 17:27:02.564: g_error_free: assertion 'error != NULL' failed
'' 
 'Accepted ' 
  @QSTRING:SSH.AUTH_METHOD@ '' 
   ' for' 
    @QSTRING:SSH_USERNAME@ '' 
     'from\ ' 
      @QSTRING:SSH_CLIENT_ADDRESS@ '' 
       'port ' 
        @NUMBER:SSH_PORT_NUMBER@ '' 
         ' ssh2' rule_id='182437592347598'
```

**Comment by jeffsf** [comment]
As amended to provide the program name

```
<!-- https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/
     3.36/administration-guide/95#TOPIC-1768868 

Identical to

     https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/
     3.26/administration-guide/81#TOPIC-1431219
-->

<patterndb version='4' pub_date='2010-10-17'>
    <ruleset name='ssh' id='123456678'>
        <pattern>ssh</pattern>
            <rules>
                <rule provider='me' id='182437592347598' class='system'>
                    <patterns>
                        <pattern>Accepted @QSTRING:SSH.AUTH_METHOD: @ for@QSTRING:SSH_USERNAME: @from\ @QSTRING:SSH_CLIENT_ADDRESS: @port @NUMBER:SSH_PORT_NUMBER:@ ssh2</pattern>
                    </patterns>
                    <examples>
                        <example>
                            <test_message program="ssh">Accepted password for sampleuser from 10.50.0.247 port 42156 ssh2</test_message>
                            <test_values>
                                <test_value name="SSH.AUTH_METHOD">password</test_value>
                                <test_value name="SSH_USERNAME">sampleuser</test_value>
                                <test_value name="SSH_CLIENT_ADDRESS">10.50.0.247</test_value>
                                <test_value name="SSH_PORT_NUMBER">42156</test_value>
                            </test_values>
                       </example>
                    </examples>
                </rule>
            </rules>
    </ruleset>
</patterndb>
```

**Comment by OverOrion** [comment]
Hi @jeffsf

Thank you for opening this issue, I can confirm that this bug happens on 3.28.1.
I've tried it with 3.36.1 and it works as intended, there is no `assertion_error`.
I believe that 6156962c4596f63bd3b23ceaf9da19e4dcbf66d1 fixed it, which should be in 3.29.1 and later releases.

I recommend using the latest release, which is currently 3.36.1.
You can download it and build it yourself or you can use the APT repository:
https://github.com/syslog-ng/syslog-ng#debianubuntu

**Comment by bazsi** [comment]
This seems to be resolved. Please reopen if you disagree.
